File: mod/Nova.Core/Communications/Endpoint.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Nova.Core.Components;
using Nova.Core.Components.Communications;
using Nova.Core.Utils;

namespace Nova.Core.Communications;

// One node in the communications graph. Aggregates one or more
// antennas and exposes a UT→position function. Vessels and ground
// stations both project to Endpoints; the Network does not know
// which is which.
public class Endpoint {

  // Stable identity. For vessels: the KSP vessel id (Guid string).
  // For ground stations: a fixed name like "KSC".
  public string Id;

  // World-frame position at the given UT. Drivers wrap the
  // appropriate KSP/Planetarium calls; tests pass deterministic stubs.
  // Authoritative for any caller that needs a numeric position.
  public Func<double, Vec3d> PositionAt;

  // Optional solver hint: structured motion model the analytical
  // horizon solver can inspect. When both endpoints of a directed
  // link expose a compatible model (e.g. both KeplerMotion around
  // the same parent body), ComputeLinkHorizons routes through
  // AnalyticalLinkHorizon instead of numerical bisection. Null →
  // always falls back to numerical.
  public MotionModel Motion;

  // True iff PositionAt is reliable for forecasting future state —
  // i.e. the closure can be queried at arbitrary future UTs and
  // returns positions consistent with what the endpoint will
  // actually do. Default true (preserves test fixture behaviour).
  //
  // Off-rails KSP vessels under thrust set this false: KSP's
  // `getTruePositionAtUT` projects a free-flight trajectory the
  // actual vessel diverges from, so any horizon computed from it
  // is wrong. ComputeLinkHorizons skips bisection for any pair
  // involving an unpredictable endpoint and pins NextEventUT at
  // the horizon cap; the driver is expected to handle bucket
  // transitions reactively (see AnyLinkBucketDifference).
  public bool IsPredictable = true;

  // Body whose SOI this endpoint currently sits in. Drives the link's
  // occluder set together with the other endpoint's PrimaryBody. May
  // be null for test fixtures without body context — such endpoints
  // behave as "always unblocked" (empty occluder set), which is the
  // safe default. The KSP-side addon sets this from
  // `vessel.orbit.referenceBody` for vessels and the home body for
  // ground stations.
  public Body PrimaryBody;

  // Antennas attached to this endpoint. The Network selects the best
  // (transmit, receive) pair per directed edge when computing rate.
  public List<Antenna> Antennas = new();

  // Cached summary of this endpoint's path back to the network's
  // designated home (KSC in the flight scene). Populated by
  // CommunicationsNetwork.RefreshHomePathSummaries after each Solve;
  // consumers (telemetry topics, UI) read these fields directly
  // instead of re-running MaxRatePath.Find every frame — the search
  // allocates a Dictionary/HashSet/List per call, which adds up at
  // 60 Hz × N vessels and shows up as GC pressure. `HasPath` is false
  // when no positive-rate route exists; on the home endpoint itself
  // the summary is `default` (self-link not meaningful).
  public PathSummary PathToHome;

  // Build an Endpoint from a VirtualVessel by harvesting its Antenna
  // components. Used by the future addon driver; useful for any
  // integration test that wants a vessel-shaped endpoint.
  public static Endpoint FromVessel(VirtualVessel vessel, string id, Func<double, Vec3d> positionAt) {
    var ep = new Endpoint { Id = id, PositionAt = positionAt };
    ep.Antennas.AddRange(vessel.AllComponents().OfType<Antenna>());
    return ep;
  }
}

// Cached connectivity summary from one endpoint to a designated home.
// Refreshed by CommunicationsNetwork once per Solve; readers can
// poll without allocating. `Direct*` fields cover the single direct
// edge (vessel→home) — they're 0 when no direct edge exists (vessel
// reachable only via relay), in which case `HasPath`/`BottleneckBps`
// still reflect the relayed path. `NextHopId` is the Id of the
// first endpoint along the chosen vessel→home path: equals home.Id
// when direct, equals a relay vessel's Id otherwise — empty when
// HasPath is false.
// `DirectSnrFloor` is the linear SNR threshold below which the
// direct edge's quantised rate drops to zero — the noise floor for
// THIS antenna pair, not the model-wide N0 = 1.0. Computed from the
// chosen TX antenna's reference SNR via the bucket-1 cutoff
// (1 + SNR_ref)^(1/N) − 1.
public struct PathSummary {
  public bool HasPath;
  public double BottleneckBps;
  public double DirectSnr;
  public double DirectRateBps;
  public double DirectMaxRateBps;
  public double DirectSnrFloor;
  public string NextHopId;
  // Ordered Links along the chosen path source→home, populated once
  // per Solve by RefreshHomePathSummaries. Null on the home endpoint
  // and when HasPath is false. Readers must treat this as a snapshot
  // — the underlying Link references are reused across Solves but
  // their fields (RateBps etc.) are stable within a Solve cycle.
  public IReadOnlyList<Link> Path;
}

File: crates/nova-sim/src/comms/endpoint.rs
//! Endpoint — a node in the comms graph. Aggregates one or more
//! antennas and resolves its world-frame position at any UT through
//! `World`. Mirrors `mod/Nova.Core/Communications/Endpoint.cs`, but
//! data-driven (no closure stored; `position_at` dispatches on
//! `EndpointKind`) so endpoints stay `Clone + Debug` across solver
//! mutations.

use crate::ephem::BodyId;
use crate::math::Vec3d;
use crate::world::{VesselId, World};

use super::antenna::Antenna;
use super::motion::{surface_offset, MotionModel};

/// Stable identity used as cache key in `(from, to)` pairs. Vessels
/// reuse their `VesselId`; ground stations get a monotonic id assigned
/// by `CommsSystem::add_ground_station`. `Ord` is derived (declaration
/// order: `Vessel < Ground`); used only to form a deterministic
/// unordered-pair key for caches.
#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum EndpointId {
    Vessel(VesselId),
    Ground(u32),
}

/// Where this endpoint's position comes from. `Vessel` resolves
/// through `World::vessel_position_absolute`; `Ground` is a fixed
/// surface point on a body, rotating with the body's spin.
#[derive(Clone, Debug)]
pub enum EndpointKind {
    Vessel(VesselId),
    Ground {
        primary: BodyId,
        latitude_deg: f64,
        longitude_deg: f64,
        altitude_m: f64,
    },
}

/// Cached connectivity summary from this endpoint to a designated
/// home. Refreshed once per Solve by
/// `CommsSystem::refresh_home_path_summaries`. Defaults to
/// no-path/zero — `default()` is the "before any solve" state and the
/// "this is the home endpoint" state.
#[derive(Copy, Clone, Debug, Default, PartialEq)]
pub struct PathSummary {
    pub has_path: bool,
    pub bottleneck_bps: f64,
    pub direct_snr: f64,
    pub direct_rate_bps: f64,
    pub direct_max_rate_bps: f64,
}

#[derive(Clone, Debug)]
pub struct Endpoint {
    pub id: EndpointId,
    pub name: String,
    pub kind: EndpointKind,
    /// Optional analytical motion hint. When both endpoints of a
    /// directed link expose a compatible model, the horizon solver
    /// runs in closed form. `None` falls back to numerical bisection
    /// on the `position_at` evaluator.
    pub motion: Option<MotionModel>,
    /// Body whose SOI this endpoint sits in. Drives the link's
    /// occluder set with the other endpoint's `primary_body`. `None`
    /// → empty occluder set (always-unblocked default for fixtures).
    pub primary_body: Option<BodyId>,
    pub antennas: Vec<Antenna>,
    /// True iff the position evaluator is reliable for forecasting.
    /// Off-rails KSP vessels under thrust set this false; the horizon
    /// solver pins their links at the horizon cap and the driver
    /// handles bucket transitions reactively. M6 has no off-rails
    /// surface — always `true`.
    pub is_predictable: bool,
    pub path_to_home: PathSummary,
}

impl Endpoint {
    /// World-frame position at `ut`. Vessels look up their orbital
    /// state via `World`; ground stations rotate with their parent
    /// body's spin.
    pub fn position_at(&self, world: &World, ut: f64) -> Vec3d {
        match &self.kind {
            // `synthesise_vessel_endpoints` skips abstract vessels, so
            // every Vessel endpoint has a positionable counterpart.
            EndpointKind::Vessel(vid) => world
                .vessel_position_absolute(*vid, ut)
                .expect("vessel endpoint synthesised for abstract vessel"),
            EndpointKind::Ground { primary, latitude_deg, longitude_deg, altitude_m } => {
                let body = world.ephemeris.body(*primary);
                let local = surface_offset(
                    body.radius,
                    body.rotation.period_seconds,
                    body.rotation.initial_rotation_rad,
                    *latitude_deg,
                    *longitude_deg,
                    *altitude_m,
                    ut,
                );
                world.ephemeris.body_position_absolute(*primary, ut) + local
            }
        }
    }
}
